### DISCUSSION TITLE
Treat GPS data like date/time

### DISCUSSION THREAD

**Discussion by Jmuccigr** [RhetTbull/osxphotos#1170] [Ideas]
One thing I often have to do is fix or add location data to my photos. As you well know, you can add this data via Photos sometime handy map in the Info box, but it doesn't get added to the original photo. It will get added if you export correctly. So it's basically like the date/time/zone data.

To get around this I've got [a few applescripts](https://github.com/Jmuccigr/AppleScripts/tree/master/Photos) that use exiftool to add GPS data to the original image file. One uses the GPS data in Photos (added via the map) and the other uses the coordinates on the clipboard (usually obtained through Apple or Google maps). Once the data get written to the original file, I invoke "Revert to original location" option in the menu, so Photos knows that it's got the right data.

So how do you feel about making a parallel command to timewarp (spacewarp?) that compares the Photos location data to the exif location data and allows a push to exif and also a writing of location data?

**Comment by RhetTbull** [comment]
It wouldn't be too hard to create a `push-exif` command that would selectively push any of the metadata to the original files. This was my original motivation for writing osxphotos 5 years ago (see [photosmeta](https://github.com/RhetTbull/photosmeta) which is no longer maintained). I abandoned this approach because I use iCloud and the changes don't get pushed to iCloud this way.  I decided it was better to have an archival copy of the photo outside the Apple ecosystem and thus `osxphotos export --exiftool` was born.  For those that don't use iCloud this would still be useful and could be implemented as a `osxphotos push-exif` command that used the same template system to set keywords, etc. You could then selectively choose which metadata to push. See #160 which suggested `osxphotos inplace` for this but I think `push-exif` might be more descriptive. 

As to your specific use case, I don't quite understand the workflow. I get changing the location data to the correct data but why the "Revert to original location" after? Does this cause Photos to re-read the data from the EXIF you just wrote? If so what advantage does this offer over just leaving your revised data in the database?

Do you use iCloud? If so I'm fairly certain the modified originals won't get pushed to iCloud which means if you ever need to re-sync your library or use another device you'll lose the modified EXIF.  I've not found any way to "force" Photos to update the original in iCloud. You could achieve something similar by editing every photo (even if just to update the EXIF data) but edited images have to be JPEG and you'd effectively double the size of your library on disk/iCloud.

If I can better understand your workflow I'll think about how to implement something like this.

**Comment by Jmuccigr** [comment]
Thanks for the reply.

Overall I want my original files to have the correct information. I like the Photos interface for downloading photos and organizing them, but I don't want the data locked into that database without being in the actual image files. So when I use time-warp, I always push to exif as well. This is just the same thing, but with GPS data instead of date/time.

To your questions, no, I don't use iCloud. I had a bad experience with my wife using it and haven't wanted to complicate things. Plus I have a lot of photos and don't want to pay any more than I am now for extra storage.

"Revert to original location" seems to do exactly as you suggest: it pulls the exif GPS data from the original file, overwriting whatever Photos had.

Basically my workflow is 

1. Get the correct GPS info. 
2. Put it into the original file.
3. Pull that GPS info into Photos.

In detail, it's two-fold at the start in order to get the correct GPS data from an image:

1a. Get the correct GPS settings for an image by looking the location up on Google or Apple maps. (This is a lat/long pair in digital form: 34.1234, 72.123.) 
1b. Use the Photos Get Info box map to drag the little stickpin to the correct location.

Now the flows merge:

2. Take that correct GPS data (either from the clipboard or from Photos) and write it to the original file via exiftool.
3. Now tell Photos to load the GPS data from the original file via the "Revert to original location" menu item. That item seems to be activated when Photos realizes that the original file and the copy it has don't have the same GPS data.

Now the original and Photos versions have the same exif data.

**Reply by RhetTbull** [reply]
That makes sense if you're not using iCloud. I'll implement this as a new osxphotos command that can push GPS info or any other metadata to the EXIF data.  

If you've pushed the Photos data to the EXIF data, they'll match so why the "revert to original" step?

**Reply by Jmuccigr** [reply]
> If you've pushed the Photos data to the EXIF data, they'll match so why the "revert to original" step?

Because Photos doesn't seem to know they match and leaves that menu option active. I imagine that it records a change in the location when I mess around in Photos and assumes they're different from what's in the exif without checking.

But I don't always push the Photos data to exif because the Photos data is wrong or missing. For example, I take a photo and forget to turn on the GPS on my camera. It's usually a lot easier and more precise to use Google or Apple maps to find the coords of where the photo was taken, so I do that. Then I use my applescript to push the coords to exif and then pull that from Photos.

**Reply by RhetTbull** [reply]
I've put together some design notes in this feature at #160

**Comment by RhetTbull** [comment]
@Jmuccigr [v0.63.0](https://github.com/RhetTbull/osxphotos/releases/tag/v0.63.0) adds a new `osxphotos push-exif` command that implements this. You can choose to "push" (write from Photos to the original files) all or any combination of metadata.  If you add `--compare` it will compare metadata and print a report of the differences between Photos and the originals without actually changing anything.

**Reply by RhetTbull** [reply]
See help for the new command [here](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-push-exif)
